File: app/TumbleKit/Model/PhotoDay.swift
import Foundation

/// A computed daily collection of prints. The photos remain stored as individual
/// SwiftData rows; this helper only shapes them for Drawer and archive screens.
public struct PhotoDay: Identifiable {
    public let dayStart: Date
    public let displayTitle: String
    public let photos: [Photo]

    public var id: Date { dayStart }
    public var totalCount: Int { photos.count }
    public var developedCount: Int { photos.filter(\.isDeveloped).count }

    public init(
        dayStart: Date,
        photos: [Photo],
        calendar: Calendar = .current,
        now: Date = Date()
    ) {
        self.dayStart = dayStart
        self.displayTitle = Self.title(for: dayStart, calendar: calendar, now: now)
        self.photos = photos.sorted { lhs, rhs in
            lhs.capturedAt > rhs.capturedAt
        }
    }

    public static func group(
        _ photos: [Photo],
        calendar: Calendar = .current,
        now: Date = Date()
    ) -> [PhotoDay] {
        Dictionary(grouping: photos) { photo in
            calendar.startOfDay(for: photo.capturedAt)
        }
        .map { dayStart, photos in
            PhotoDay(dayStart: dayStart, photos: photos, calendar: calendar, now: now)
        }
        .sorted { lhs, rhs in
            lhs.dayStart > rhs.dayStart
        }
    }

    private static func title(for dayStart: Date, calendar: Calendar, now: Date) -> String {
        if calendar.isDate(dayStart, inSameDayAs: now) {
            return "Today"
        }

        if let yesterday = calendar.date(byAdding: .day, value: -1, to: now),
           calendar.isDate(dayStart, inSameDayAs: yesterday) {
            return "Yesterday"
        }

        return dayStart.formatted(.dateTime.month(.abbreviated).day().year())
    }
}

File: android/app/src/main/java/com/tumble/model/PhotoDay.kt
package com.tumble.model

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * A computed daily collection of prints. The photos remain stored as individual
 * Room rows; this helper only shapes them for Drawer and archive screens.
 *
 * Ported from `app/TumbleKit/Model/PhotoDay.swift`.
 */
data class PhotoDay(
    val dayStart: LocalDate,
    val displayTitle: String,
    val photos: List<Photo>,
) {
    val totalCount: Int get() = photos.size
    val developedCount: Int get() = photos.count { it.isDeveloped }

    companion object {
        private val fallbackFormat: DateTimeFormatter =
            DateTimeFormatter.ofPattern("MMM d, yyyy", Locale.US)

        fun group(
            photos: List<Photo>,
            zone: ZoneId = ZoneId.systemDefault(),
            now: Instant = Instant.now(),
        ): List<PhotoDay> =
            photos
                .groupBy { it.capturedAt.atZone(zone).toLocalDate() }
                .map { (dayStart, dayPhotos) ->
                    PhotoDay(
                        dayStart = dayStart,
                        displayTitle = title(dayStart, zone, now),
                        photos = dayPhotos.sortedByDescending { it.capturedAt },
                    )
                }
                .sortedByDescending { it.dayStart }

        private fun title(dayStart: LocalDate, zone: ZoneId, now: Instant): String {
            val today = now.atZone(zone).toLocalDate()
            return when (dayStart) {
                today -> "Today"
                today.minusDays(1) -> "Yesterday"
                else -> dayStart.format(fallbackFormat)
            }
        }
    }
}
